kiwi/lib/search.py

```diff
@@ -0,0 +1,655 @@
+                    'dropout'
+                ] = dropout
+            search_values['dropout'] = dropout
+        # Suggest the number of warmup steps
+        if self.config.options.warmup_steps is not None:
+            warmup_steps = get_suggestion(
+                trial, 'warmup_steps', self.config.options.warmup_steps
+            )
+            base_config_dict['system']['optimizer']['warmup_steps'] = warmup_steps
+            search_values['warmup_steps'] = warmup_steps
+        # Suggest the number of freeze epochs
+        if self.config.options.freeze_epochs is not None:
+            freeze_epochs = get_suggestion(
+                trial, 'freeze_epochs', self.config.options.freeze_epochs
+            )
+            base_config_dict['system']['model']['encoder']['freeze'] = True
+            base_config_dict['system']['model']['encoder'][
+                'freeze_for_number_of_steps'
+            ] = int(self.updates_per_epochs * freeze_epochs)
+            search_values['freeze_epochs'] = freeze_epochs
+        # Suggest a hidden size
+        if self.config.options.hidden_size is not None:
+            hidden_size = get_suggestion(
+                trial, 'hidden_size', self.config.options.hidden_size
+            )
+            base_config_dict['system']['model']['encoder']['hidden_size'] = hidden_size
+            base_config_dict['system']['model']['decoder']['hidden_size'] = hidden_size
+            search_values['hidden_size'] = hidden_size
+        # Suggest a bottleneck size
+        if self.config.options.bottleneck_size is not None:
+            bottleneck_size = get_suggestion(
+                trial, 'bottleneck_size', self.config.options.bottleneck_size
+            )
+            base_config_dict['system']['model']['decoder'][
+                'bottleneck_size'
+            ] = bottleneck_size
+            search_values['bottleneck_size'] = bottleneck_size
+        # Suggest whether to use the MLP after the encoder
+        if self.config.options.search_mlp:
+            use_mlp = trial.suggest_categorical('mlp', [True, False])
+            base_config_dict['system']['model']['encoder']['use_mlp'] = use_mlp
+            search_values['use_mlp'] = use_mlp
+
+        # Search word_level and sentence_level and their combinations
+        # Suggest whether to include the sentence level objective
+        if self.config.options.search_hter:
+            assert (
+                base_config_dict['data']['train']['output']['sentence_scores']
+                is not None
+            )
+            assert (
+                base_config_dict['data']['valid']['output']['sentence_scores']
+                is not None
+            )
+            hter = trial.suggest_categorical('hter', [True, False])
+            base_config_dict['system']['model']['outputs']['sentence_level'][
+                'hter'
+            ] = hter
+            search_values['hter'] = hter
```

Comment:
minor missing a newline

Reply:
I don't get this one. What is missing a newline?